controllers/topolvm_controller.go

```diff
@@ -0,0 +1,259 @@
+package controllers
+
+import (
+	"context"
+
+	lvmv1alpha1 "github.com/red-hat-storage/lvm-operator/api/v1alpha1"
+	appsv1 "k8s.io/api/apps/v1"
+	v1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	cutil "sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+)
+
+const (
+	controllerName = "topolvm-controller"
+)
+
+type controllerPlugin struct{}
+
+// controllerPlugin unit satisfies resourceManager interface
+var _ resourceManager = controllerPlugin{}
+
+func (c controllerPlugin) getName() string {
+	return controllerName
+}
+
+func (c controllerPlugin) ensureCreated(r *LVMClusterReconciler, ctx context.Context, lc *lvmv1alpha1.LVMCluster) error {
+
+	// get the desired state of topolvm controller deployment
+	cd := getControllerDeployment(lc)
+	result, err := cutil.CreateOrUpdate(ctx, r.Client, cd, func() error {
+		// make sure LVMCluster CR garbage collects controller deployments and also block owner removal
+		return cutil.SetControllerReference(lc, cd, r.Scheme)
+	})
+
+	// log the result based on current reconcile
+	switch result {
+	case cutil.OperationResultCreated:
+		r.Log.Info("csi controller", "operation", result, "name", cd.Name)
+	case cutil.OperationResultUpdated:
+		r.Log.Info("csi controller", "operation", result, "name", cd.Name)
+	case cutil.OperationResultNone:
+		r.Log.Info("csi controller", "operation", result, "name", cd.Name)
+	default:
+		r.Log.Error(err, "csi controller reconcile failure", "name", cd.Name)
+		return err
+	}
+
+	return nil
+}
+
+func (c controllerPlugin) ensureDeleted(r *LVMClusterReconciler, ctx context.Context, lc *lvmv1alpha1.LVMCluster) error {
+	cd := &appsv1.Deployment{}
+	err := r.Client.Get(ctx, types.NamespacedName{Name: TopolvmControllerDeploymentName, Namespace: lc.Namespace}, cd)
+
+	if err != nil {
+		// deployment is already deleted based on ownerref and just log the msg
+		if errors.IsNotFound(err) {
+			r.Log.Info("csi controller deleted", "TopolvmController", cd.Name)
+			return nil
+		}
+		r.Log.Error(err, "unable to retrieve csi controller deployment", "TopolvmController", cd.Name)
+		return err
+	}
+	return nil
```

Comment:
The deployment should be deleted here.

Reply:
For a deployment, an ownerReference should be enough for cleanup.
Not for cluster scoped objects or complex cleanups though.

Reply:
What is the expected behaviour if the lvmcluster object is deleted? Shouldn't the deployment be deleted as well?

Reply:
- I understood that if lvm cr is deleted then controller will also be deleted due to owner ref
- However I'm unsure whether dependents will be deleted after realizing deletestamp on owner or owner needs to be deleted (like actual deletion ~Not found)🤔

Reply:
At what point will the deployment be deleted? As soon as the delete operation is performed or after the finalizer is removed from the LvmController? The operator waits for all resources to be deleted before removing the finalizer.

If it is deleted before the finalizer is removed, then this method should return an error as long as  the deployment resource exists in order to prevent the removal of the finalizer. If not, this is not a good candidate for the resource manager.

Reply:
@nbalacha 

- I did some testing using configmaps with finalizers and ownerreferences and here's the gist in current context

1. If controller deployment were to depend on lvmcluster CR, if I set non-blocking ownerref on deployment and you set finalizer on CR, when you remove the finalizer then only controller deployment will be deleted
2. However the current reconciler only removes the finalizer when all the units get deleted, in this case I do not depend on lvm CR to be the owner and listen to reconciler to delete deployment whenever it's called

- I see two ways out of it
1. You do not change the reconcile and send delete op to the resource units and I'll add deletion logic
2. You can just remove the finalizer and send a delete op to the resource units and reconcile for nil (because some resources like csidriver is cluster scoped and doesn't respect ownerref)?

Reply:
Lets stick with the current approach for now - the ensureDeleted will effectively be a noop for topolvm-controller. The lvmcluster controller will remove the finalizer when all resource managers return nil . The LVMCluster object will then be removed followed the topolvm-controller.

Reply:
- uh oh, just saw this review after adding tests 
- anyways will raise separate PRs

Reply:
- I'm continuing with a non-blocking owner reference and also implementing custom logic for controller deletion
- We can decide what exactly to be used as it's not of utmost importance as of now